### PR TITLE
[BEAM-13459] Cache artifacts across job runs on python+dataflow

### PR DESCRIPTION
**Pull request by steveniemitz**
This adds support for caching artifacts across job runs to the python dataflow runner.  The method it uses is slightly different than the JVM uses to work around some limitations to the python container boot code.  Specifically:

- Certain artifacts must have a specific name (dataflow_python_sdk.tar, pipeline.pb, etc)
- All artifacts for a job must be in the same "directory" on GCS

To work around this, we stage artifacts in a two step process now (gated by a pipeline flag, default value TBD?):
- First upload the artifact to a common cache location (`artifact_cache`) directly under the staging directory.  In this location the artifact is named using its sha256 hash.  If the file already exists, this step is skipped (which is the slowest step)
- Once uploaded (either because it was a cache hit, or just uploaded), perform a GCS copy operation to "copy" the file from its cached location into where it should be in the job's unique staging directory.  This operation is fast since its a GCS->GCS copy in the same bucket.

Ideally we wouldn't need these workarounds, but this works well for now.  If the boot code is ever enhanced to fix the above two issues, it'd be easy to change this to use a new caching method.

Just choosing some reviewers based on who's last changed code here:

R: @ihji @tvalentyn 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by steveniemitz** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
Thanks a lot @steveniemitz, haven't taken a closer look yet, but cc: @AnandInguva who is also looking in the pipeline dependency staging functionality.

**Comment by AnandInguva** [comment]
@tvalentyn a quick question. Do we delete artifacts that are cached after a certain period (eg: 1 month)?

**Comment by steveniemitz** [comment]
Users can set expiration policies on their buckets if they want, but there's nothing built-in.  Particularly with streaming jobs expiration can be dangerous because the job can always outlive the artifact expiration, if that happens the job will fail the next time it needs an artifact (autoscaling, worker failure, etc).

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py [hunk 1]
```diff
@@ -1385,6 +1388,246 @@ def test_enable_hot_key_logging(self):
...
+    return [
+        mock.call.gcs_exists(gcs_cache_path),
+        mock.call.gcs_upload(src, gcs_cache_path),
+        mock.call.gcs_gcs_copy(
+            gcs_cache_path, f'gs://test-location/staging/{dst_name}')
+    ]
+
+  def _mock_cached_copy(self, staging_root, src, sha256, dst_name=None):
+    uncached = self._mock_uncached_copy(staging_root, src, sha256, dst_name)
+    uncached.pop(1)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py:1410]: I am having trouble following the logic in this unit test.
As an example, could you please explain what happens in this helper?
Is the intent for helpers `_mock_uncached_copy`, `_mock_cached_copy` to return expected calls for an upload that had a cache hit and an upload that had a cache miss? There may be name conflation with helpers in the code `_uncached_gcs_file_copy` (upload without caching enabled) `_cached_gcs_file_copy` (upload without  caching enabled)

**Review by tvalentyn — COMMENTED** [review]
Thanks!

**Comment by tvalentyn** [comment]
> All artifacts for a job must be in the same "directory" on GCS

For my eductaion, where is this requirement coming from?

**Comment by steveniemitz** [comment]
> > All artifacts for a job must be in the same "directory" on GCS
> 
> For my eductaion, where is this requirement coming from?

The python container boot code (the go binary) ensures it.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 2]
```diff
@@ -556,9 +565,79 @@ def _get_sdk_image_overrides(self, pipeline_options):
...
+      request = storage.StorageObjectsGetRequest(bucket=bucket, object=name)
+      try:
+        self._storage_client.objects.Get(request)
+        return True
+      except exceptions.HttpError as e:
+        return e.status_code not in (403, 404)
+
+  @retry.with_exponential_backoff(
+      retry_filter=retry.retry_on_server_errors_and_timeout_filter)
+  def _gcs_to_gcs_copy(self, from_gcs, to_gcs):
```
[the remark this answers is not in the document]
Comment (steveniemitz) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:608]: I didn't try that (or even knew it existed), it seems like it'd indeed remove duplication here if it works correctly.

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 3]
```diff
@@ -715,6 +715,13 @@ def _add_argparse_args(cls, parser):
         help='When true, will enable the direct logging of any detected hot '
         'keys into Cloud Logging. Warning: this will log the literal key as an '
         'unobfuscated string.')
+    parser.add_argument(
+        '--enable_artifact_caching',
+        default=False,
```
[the remark this answers is not in the document]
Comment (steveniemitz) [reply sdks/python/apache_beam/options/pipeline_options.py:720]: no, its always on, there's no way to disable it in java.  As I mentioned in the comment, I don't have any particular strong opinion on if its on or off by default here.  I'd prefer its on by default but if that's a contentious subject I'll just leave it off.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 4]
```diff
@@ -519,11 +523,16 @@ def __reduce__(self):
 
 
 class DataflowApplicationClient(object):
+  _HASH_CHUNK_SIZE = 1024 * 8
+  _GCS_CACHE_PREFIX = "artifact_cache"
   """A Dataflow API client used by application code to create and query jobs."""
-  def __init__(self, options):
+  def __init__(self, options, root_staging_location=None):
```
[the remark this answers is not in the document]
Comment (steveniemitz) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:529]: it's used above https://github.com/apache/beam/pull/16229/files#diff-d7e8ff5775f3fb328137483bd443e04d96caa92f3f68a188391f18a8e06075a0R574

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 5]
```diff
@@ -556,9 +565,79 @@ def _get_sdk_image_overrides(self, pipeline_options):
...
+    if not gcs_or_local_path.startswith('gs://'):
+      return False
+    else:
+      bucket, name = self._split_gcs_path(gcs_or_local_path)
+      request = storage.StorageObjectsGetRequest(bucket=bucket, object=name)
+      try:
+        self._storage_client.objects.Get(request)
+        return True
+      except exceptions.HttpError as e:
+        return e.status_code not in (403, 404)
```
[the remark this answers is not in the document]
Comment (steveniemitz) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:604]: yeah that seems right.  to your point about copy though, this should probably just use FileSystems.exists

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py (hunk 1 above)
Comment (steveniemitz) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py:1410]: the difference between a cached and uncached copy (from a mocked call POV) is simply that the cached copy doesn't do an upload to GCS to the `gcs_cache_path`, this simply uses the expected calls generated by `_mock_uncached_copy` and removes the "upload" call.

**Comment by tvalentyn** [comment]
Thanks. Please don't squash reviewed and unreviewed commits before the review has completed, see: https://beam.apache.org/contribute/#make-the-reviewers-job-easier.

**Comment by steveniemitz** [comment]
It's always a gamble that the person doing the merge will actually squash it :P

File: sdks/python/apache_beam/options/pipeline_options.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/options/pipeline_options.py:720]: We can have default as false for one release, announce the flag in CHANGES.MD then make it true in one of the subsequent releases module any received feedback.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py:1411]: thanks, i misread this code thinking that it returned the result of `uncached.pop(1)`

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
Looks like this change makes it into 2.36.0 release cut, please update CHANGES.MD

**Pull request merged**